### PR TITLE
Tests rimjoists vs plenums

### PR DESCRIPTION
**Pull request by brgix**
Carries over this extended [test](https://github.com/rd2/tbd_tests/pull/27) dealing with indirectly-conditioned spaces (like return air plenums) and how perimeter edges of their "floors" (i.e. suspended ceiling tiles) are (often) tagged as "rimjoists" (i.e. intermediate slab edges) ... something to remediate (the test is a first step).

### CODE REVIEW

File: lib/tbd/geo.rb [hunk 1]
```diff
@@ -347,6 +347,7 @@ def properties(surface = nil, argh = {})
 
     surf[:conditioned] = surf.key?(:heating) || surf.key?(:cooling)
     surf[:space      ] = space
+    surf[:occupied   ] = space.partofTotalFloorArea
```
Comment (brgix) [inline lib/tbd/geo.rb:350]: Keeping track of _in/directly-conditioned_ spaces that are 'unoccupied' (i.e plenums).

File: spec/tbd_tests_spec.rb [hunk 2]
```diff
@@ -2039,9 +2040,11 @@
     expect(model).to_not be_empty
     model = model.get
 
+    # Ensure the plenum is 'unoccupied', i.e. not part of the total floor area.
     plnum = model.getSpaceByName("PLENUM-1")
     expect(plnum).to_not be_empty
     plnum = plnum.get
+    expect(plnum.setPartofTotalFloorArea(false)).to be true
```
Comment (brgix) [inline spec/tbd_tests_spec.rb:2047]: Key in having TBD correctly identify _in/directly-conditioned_ spaces as 'unoccupied'.

File: spec/tbd_tests_spec.rb [hunk 3]
```diff
@@ -2080,39 +2083,43 @@
       expect(surfaces[s][:deratable]).to be false
     end
 
-    # There are now above-grade "rimjoists", i.e. edge along suspended ceilings:
-    expect(io[:edges].count { |edge| edge[:type] == :rimjoist      }).to eq(4)
+    # Prior to v3.4.0, plenum floors would have been tagged as "rimjoists". No
+    # longer the case ("ceilings" are caught earlier in the process).
+    expect(io[:edges].count { |edge| edge[:type] == :ceiling       }).to eq(4)
+    expect(io[:edges].count { |edge| edge[:type] == :rimjoist      }).to eq(0)
```
Comment (brgix) [inline spec/tbd_tests_spec.rb:2089]: Now working as expected: no return air plenum floors as "rimjoists", just (benign) "ceilings".

File: spec/tbd_tests_spec.rb [hunk 4]
```diff
@@ -2080,39 +2083,43 @@
...
+    # users could always override this default behaviour by specifying spacetype
+    # -specific PSI factor sets (JSON inputs), with "rimjoists" of 0 W/K per
+    # meter. Yet these workarounds necessarily implied additional steps for the
+    # vast majority of TBD users. As of v3.4.0, the default automated TBD
+    # outcome is to tag plenum "floors" as "ceilings" (no additional steps).
     #
-    # In hindsight, these workarounds imply additional steps for the vast
-    # majority of TBD users. The previous steps illustrate the current situation,
-    # which constitute the first step towards a revised solution (soonish).
+    # The flip side is that additional consideration may be required for less
```
Comment (brgix) [inline spec/tbd_tests_spec.rb:2111]: These nuances should probably make it to the online User Guide ...

File: tbd.schema.json [hunk 5]
```diff
@@ -122,6 +122,18 @@
...
+          "type": "number"
+        },
+        "ceilingconcave": {
+          "title": "Concave intermediate ceiling (not floor) edge PSI",
+          "type": "number"
+        },
+        "ceilingconvex": {
+          "title": "Convex intermediate ceiling (not floor) edge PSI",
+          "type": "number"
+        },
```
Comment (brgix) [inline tbd.schema.json:136]: Needs additional tests, e.g.:

- how to customize "ceiling" PSI factors
- how to override "ceiling" PSI factors as e.g. "rimjoists"

TO DO ...

File: lib/tbd/psi.rb [hunk 6]
```diff
@@ -156,6 +156,7 @@ def initialize
         rimjoist:        1.000000, # re: BETBG
         parapet:         0.800000, # re: BETBG
         roof:            0.800000, # same as parapet
+        ceiling:         0.000000, # e.g. suspended ceiling tiles
```
Comment (brgix) [inline lib/tbd/psi.rb:159]: Adding "ceiling" PSI factors of 0 W/K per linear meter by default.

File: lib/tbd/psi.rb [hunk 7]
```diff
@@ -1866,13 +1897,20 @@ def process(model = nil, argh = {})
...
         is[:roof            ] = set.keys.to_s.include?("roof")
+        is[:ceiling         ] = set.keys.to_s.include?("ceiling")
         is[:party           ] = set.keys.to_s.include?("party")
         is[:grade           ] = set.keys.to_s.include?("grade")
         is[:balcony         ] = set.keys.to_s.include?("balcony")
         is[:balconysill     ] = set.keys.to_s.include?("balconysill")
         is[:balconydoorsill ] = set.keys.to_s.include?("balconydoorsill")
         is[:rimjoist        ] = set.keys.to_s.include?("rimjoist")
 
+        if is.empty?
```
Comment (brgix) [inline lib/tbd/psi.rb:1908]: Nothing to do with plenum floors/ceilings. Simply wiser to post-process more generic tags.

File: lib/tbd/psi.rb [hunk 8]
```diff
@@ -1866,13 +1897,20 @@ def process(model = nil, argh = {})
...
+          is[:head] = set.keys.to_s.include?("head")
+          is[:sill] = set.keys.to_s.include?("sill")
+          is[:jamb] = set.keys.to_s.include?("jamb")
+        end
```
Comment (brgix) [inline lib/tbd/psi.rb:1912]: Nothing to do with plenum floors/ceilings. Simply wiser to post-process more generic tags.

File: lib/tbd/psi.rb [hunk 9]
```diff
@@ -2092,6 +2130,41 @@ def process(model = nil, argh = {})
...
+          next      if i == id
+          next  unless floors.key?(i)
+          next      if floors[i][:ground     ]
+          next  unless floors[i][:conditioned]
+          next      if floors[i][:occupied   ]
+
+          ceiling = floors[i][:boundary]
+          next  unless ceilings.key?(ceiling)
+          next  unless ceilings[ceiling][:conditioned]
+          next  unless ceilings[ceiling][:occupied   ]
```
Comment (brgix) [inline lib/tbd/psi.rb:2150]: New loop/logic to tag edges as "ceiling".

File: lib/tbd/psi.rb [hunk 10]
```diff
@@ -2259,15 +2333,17 @@ def process(model = nil, argh = {})
         end
 
         edge[:surfaces].keys.each do |i|
-          break     if is[:rimjoist   ] || is[:balcony        ] ||
-                       is[:balconysill] || is[:balconydoorsill]
+          break     if is[:ceiling        ]
```
Comment (brgix) [inline lib/tbd/psi.rb:2336]: If an edge has been previously tagged as a "ceiling", don't bother checking for "rimjoist/balcony" variants.

File: json/tbd_z5.json [hunk 11]
```diff
@@ -0,0 +1,12 @@
+{
+  "schema": "https://github.com/rd2/tbd/blob/master/tbd.schema.json",
+  "description": "testing TBD JSON for 5Zone (plenum)",
+  "psis": [{
+      "id": "salk",
+      "ceiling": 0.1
+  }],
+  "spaces": [{
+    "id": "PLENUM-1",
+    "psi": "salk"
```
Comment (brgix) [inline json/tbd_z5.json:10]: Somewhat hypothetical experiment: A "Salk Institute"-like plenum floor case (where technical staff walk around on). Postulating here that the _structural_ plenum floor requires perimeter brackets along exterior walls (i.e. thermal bridges of 0.1 W/K per linear meter).

File: spec/tbd_tests_spec.rb [hunk 12]
```diff
@@ -2121,6 +2123,58 @@
...
+    other_ceilings.each do |s|
+      expect(surfaces).to have_key(s)
+      expect(surfaces[s][:deratable]).to be false
+    end
+
+    io[:edges].select { |edge| edge[:type] == :ceiling }.each do |edge|
+      expect(edge[:psi]).to eq("salk")
+    end
+
+    expect(io[:edges].count { |edge| edge[:type] == :ceiling       }).to eq(4)
```
Comment (brgix) [inline spec/tbd_tests_spec.rb:2169]: The "Salk Institute"-like plenum floor edges remain tagged as "ceiling", yet retain a custom PSI-factor of 0.1 W/K per linear meter (not the default 0). Recourse to JSON customization implies additional steps in such cases, but "Salk Institute"-like plenums are a rare occurrence, albeit fascinating.

File: lib/tbd/geo.rb (hunk 1 above)
Comment (macumber) [reply lib/tbd/geo.rb:350]: This seems like a good idea.  For actual plenums you can check [`isPlenum`](https://github.com/NREL/OpenStudio/blob/develop/src/model/ThermalZone.hpp#L438) or [ `canBePlenum`](https://github.com/NREL/OpenStudio/blob/develop/src/model/ThermalZone.hpp#L444C10-L444C21).  That doesn't help as much for attics or crawlspaces so I think what you have here covers everything.

Comment (brgix) [reply lib/tbd/geo.rb:350]: Thanks @macumber. In the first releases, we actually did rely on a "plenum" spacetype-based method (which in part relied on _isPlenum_). [Here](https://github.com/rd2/osut/blob/362e0ac34426987d5b49c9e16e9ed71779ccd061/lib/osut/utils.rb#L1850), I go through why relying on _plenum_ categorization is inadequate. The solution now rests on [OSut::unconditioned?](https://github.com/rd2/osut/blob/362e0ac34426987d5b49c9e16e9ed71779ccd061/lib/osut/utils.rb#L2052)(space), which calls [OSut::setpoint(space)](https://github.com/rd2/osut/blob/362e0ac34426987d5b49c9e16e9ed71779ccd061/lib/osut/utils.rb#L1961). The latter catches indirectly-conditioned spaces, so yes we're OK on that front. What was missing here was catching whether the space was occupied or not.

**Pull request merged**